openshift/machine-config-operator
# MCO-1100: enable RHEL entitlements in on-cluster layering with OCL API

**- What I did**

This adds the capability for BuildController to use the RHEL entitlement secrets to allow cluster admins to inject RHEL content into their builds that they are entitled to receive. This also allows the injection / consumption of content into `/etc/yum.repos.d` as well as `/etc/pki/rpm-gpg`. There are a few notes about the implementation that I would like to have at a higher level:

- Because we run rootless Buildah, we're more prone to running into SELinux complications. This makes it more difficult  to directly mount the contents of `/etc/yum.repos.d`, `/etc/pki/entitlement`, and `/etc/pki/rpm-gpg` directly into the build context. With that in mind, we copy everything into a series of temp directories first, and then mount those temp directories into the build context as a volume.
- We also create an `emptyDir` which is mounted into the build pod at `/home/build/.local/share/containers`. It is unclear why this is necessary, but as mentioned before, I suspect that this is due to SELinux issues.
- The e2e test suite now has the capability to stream the container logs from the build pod to the filesystem as there is useful information contained within those logs if the e2e test fails. In OpenShift CI, this location will be determined by the `ARTIFACT_DIR` env var. If this env var is not present, it will default the current directory.
- For right now, `etc-pki-entitlement` flow (specifically, the `TestEntitledBuild` test) is being skipped in OpenShift CI because the test clusters do not have that cred available. The test suite will automatically detect the presence (or lack thereof) of that cred in the `openshift-config-managed` namespace and run the test if it is available. However, the `TestYumRepos` test targets a very similar flow and can do its own setup and teardown regardless of creds preexisting.

I took care to ensure that this does not break OKD by taking the following actions:
- I observed that the addition of the `/home/build/.local/share/containers` volume mount to the build pod prevented the `wait-for-done` container to start when running on FCOS. With this in mind, I modified the build pod instantiation to not connect this volume mount to the `wait-for-done` container.
- I added a ` TestOnClusterBuildsOnOKD` e2e test which will only run against an OKD cluster. Conversely, I excluded other tests from running against an OKD cluster since those tests make assumptions about things that would only be present within an OCP cluster.

The difference between this PR and https://github.com/openshift/machine-config-operator/pull/4312 is that this one is based upon both the on-cluster layering PR (https://github.com/openshift/machine-config-operator/pull/4327) and the on-cluster layering e2e PR (https://github.com/openshift/machine-config-operator/pull/4328).

**- How to verify it**

Automated verification:

1. Bring up a cluster where the secret `etc-pki-entitlement` exists in the `openshift-config-managed` namespace. If this secret is not present, `TestEntitledBuilds` and `TestEntitledBuildsRollsOutImage` will be skipped.
2. Ensure that the OnClusterBuild feature-gate is enabled. The test suite will fail immediately if the feature-gate is not enabled.
3. Run the tech preview e2e test suite: `$ go test -count=1 -v ./test/e2e-techpreview/...`

(Note: Because we have not landed https://github.com/openshift/machine-config-operator/pull/4284, the cleanup / teardown will delete the node and its underlying machine, causing the Machine API to provision a replacement node.)

Manual verification:

1. Bring up a cluster where the secret `etc-pki-entitlement` exists in the `openshift-config-managed` namespace. 
2. Ensure that the OnClusterBuild feature-gate is enabled.
3. Copy the `etc-pki-entitlement` secret into the `openshift-machine-config-operator` namespace. Here's a small script you can use:
```bash
#!/usr/bin/env bash

set -xeuo

oc create secret generic etc-pki-entitlement \
  --namespace "openshift-machine-config-operator" \
  --from-file=entitlement.pem=<(oc get secret/etc-pki-entitlement -n openshift-config-managed -o go-template='{{index .data "entitlement.pem" | base64decode }}') \
  --from-file=entitlement-key.pem=<(oc get secret/etc-pki-entitlement -n openshift-config-managed -o go-template='{{index .data "entitlement-key.pem" | base64decode }}')
```
5. Create a new MachineConfigPool and MachineOSConfig :
```yaml
---
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfigPool
metadata:
  name: layered
spec:
  machineConfigSelector:
    matchExpressions:
    - key: machineconfiguration.openshift.io/role
      operator: In
      values:
      - worker
      - layered
  nodeSelector:
    matchLabels:
      node-role.kubernetes.io/layered: ""
  paused: false
---
apiVersion: machineconfiguration.openshift.io/v1alpha1
kind: MachineOSConfig
metadata:
  name: layered
spec:
  buildInputs:
    baseImagePullSecret:
      name: global-pull-secret-copy
    containerFile:
    - containerfileArch: noarch
      content: |-
        FROM configs AS final
        RUN rpm-ostree install cowsay && \
          ostree container commit
    imageBuilder:
      imageBuilderType: PodImageBuilder
    renderedImagePushSecret:
      name: <add your registry's push secret here>
    renderedImagePushspec: <add your registry's pushspec here>
  buildOutputs:
    currentImagePullSecret:
      name: <add your registry's pull secret here>
  machineConfigPool:
    name: layered
```
6. Watch for the `machine-os-builder` pod to start. Shortly afterward, the build pod should start. It should complete without any errors. Seeing the following lines in the build pod content will verify that we've successfully ingested content:
```console
Enabled rpm-md repositories: rhel-9-for-x86_64-baseos-beta-rpms rhel-9-for-x86_64-appstream-beta-rpms
Updating metadata for 'rhel-9-for-x86_64-baseos-beta-rpms'...done
Updating metadata for 'rhel-9-for-x86_64-appstream-beta-rpms'...done
Importing rpm-md...done
rpm-md repo 'rhel-9-for-x86_64-baseos-beta-rpms'; generated: 2024-03-25T12:33:56Z solvables: 1816
rpm-md repo 'rhel-9-for-x86_64-appstream-beta-rpms'; generated: 2024-03-25T12:34:53Z solvables: 6972
```

**- Description for the changelog**
Enables RHEL entitlements in on-cluster layering
